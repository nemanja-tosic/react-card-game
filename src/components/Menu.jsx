// @flow
import React from 'react';

import './Menu.css';

/**
 * The end number is inclusive.
 */
const range = (start: number, end: number) => {
  return Array.from({length: (end - start + 1)}, (v, k) => k + start);
};

class Menu extends React.Component {
  state = {
    playerName: '',
    numberOfOpponents: 1,
  }

  props: {
    onStartGame: (playerName: string, numberOfOpponents: number) => void,
  }

  constructor(props: any) {
    super(props);

    (this: any).handleChangeName = this.handleChangeName.bind(this);
    (this: any).handleChangeOpponent = this.handleChangeOpponent.bind(this);
    (this: any).handleSubmit = this.handleSubmit.bind(this);
  }

  handleChangeName(event: Event & { currentTarget: HTMLInputElement }): void {
    const input = event.currentTarget;
    const inputInvalidMessage = 'The name must not start with a whitespace';

    input.setCustomValidity(input.validity.patternMismatch ? inputInvalidMessage : '');
    this.setState({playerName: input.value});
  }

  handleChangeOpponent(event: Event & { currentTarget: HTMLInputElement }): void {
    this.setState({numberOfOpponents: parseInt(event.currentTarget.value, 10)});
  }

  handleSubmit(event: Event): void {
    this.props.onStartGame(this.state.playerName, this.state.numberOfOpponents);
    event.preventDefault();
  }

  render() {
    return (
      <div className="Menu container-fluid">
        <div className="row justify-content-center align-items-center">
          <form className="col-sm-6" onSubmit={this.handleSubmit}>
            <div className="form-group Menu__name">
              <input
                maxLength="20"
                pattern="\S.*"
                onChange={this.handleChangeName}
                className="form-control"
                placeholder="Your name"
                required/>
            </div>
            <div className="form-group Menu__number-of-opponents">
              <select className="form-control" onChange={this.handleChangeOpponent}>
                {range(1, 3).map(player => (
                  <option key={player} value={player}>{player}</option>
                ))}
              </select>
            </div>
            <div className="form-group">
              <button type="submit" className="btn btn-primary">Start game</button>
            </div>
          </form>
        </div>
      </div>
    );
  }
}

export default Menu;
