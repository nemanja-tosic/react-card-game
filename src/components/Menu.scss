.Menu form {
  max-width: 400px;
}

.Menu, .Menu > .row {
  height: 100%;
}
