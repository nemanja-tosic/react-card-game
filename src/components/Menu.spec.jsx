// @flow
import React from 'react';
import { shallow } from 'enzyme';

import Menu from './Menu';

describe('Menu', () => {
  let component;
  let startGameSpy;

  beforeEach(() => {
    startGameSpy = jest.fn();
    component = shallow(<Menu onStartGame={startGameSpy}/>);
  });

  it('should show a form with entries for the player name and the number of opponents', () => {
    expect(component.getNode()).toMatchSnapshot();
  });

  describe('when the start game button is clicked', () => {
    describe('and the form is valid', () => {
      it('should call the onStartGame prop with the entered details', () => {
        const name = 'my name';
        const numberOfOpponents = 3;
        const nameInput = component.find('.Menu__name input');
        const numberOfOpponentsDropdown = component.find('.Menu__number-of-opponents select');
        const form = component.find('form');

        nameInput.simulate('change', {
          currentTarget: {
            value: name,
            validity: {
              patternMismatch: false,
            },
            setCustomValidity: () => {
            }
          },
        });
        numberOfOpponentsDropdown.simulate('change', {
          currentTarget: {
            value: numberOfOpponents,
          },
        });
        form.simulate('submit', {
          preventDefault: () => {
          },
        });

        expect(startGameSpy).toHaveBeenCalledWith(name, numberOfOpponents);
      });
    });
  });
});
