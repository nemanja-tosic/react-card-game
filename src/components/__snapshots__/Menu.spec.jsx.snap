// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu should show a form with entries for the player name and the number of opponents 1`] = `
<div
  className="Menu container-fluid"
>
  <div
    className="row justify-content-center align-items-center"
  >
    <form
      className="col-sm-6"
      onSubmit={[Function]}
    >
      <div
        className="form-group Menu__name"
      >
        <input
          className="form-control"
          maxLength="20"
          onChange={[Function]}
          pattern="\\\\S.*"
          placeholder="Your name"
          required={true}
        />
      </div>
      <div
        className="form-group Menu__number-of-opponents"
      >
        <select
          className="form-control"
          onChange={[Function]}
        >
          <option
            value={1}
          >
            1
          </option>
          <option
            value={2}
          >
            2
          </option>
          <option
            value={3}
          >
            3
          </option>
        </select>
      </div>
      <div
        className="form-group"
      >
        <button
          className="btn btn-primary"
          type="submit"
        >
          Start game
        </button>
      </div>
    </form>
  </div>
</div>
`;
